### PR TITLE
[PVM, DAC] Admin proposals

### PR DESCRIPTION
**Pull request by evlekht**
## Why this should be merged
This PR does 2 things:
-  introduces feature "admin proposals"
- forbids addressStateTx to edit consortium bit

Admin proposals are proposals that can be created only by allowed role (on proposal-type basis) and will be executed immediately after creation without any voting. So, they're more like admin actions than proposals.
### Admin proposals
PR adds new proposal type that wraps other proposal types.
```go
type AdminProposal struct {
	OptionIndex uint32 `serialize:"true"`
	Proposal    `serialize:"true"`
}
```
If proposal of this type is inside of addProposalTx, than tx will try to create admin proposal.

On execution node will check if wrapped proposal type actually allows be used as admin proposal, than it will check that proposer has address state role that allows him to create admin proposals of this type.

If all checks successful, node will create proposal state with already existing vote for specified option index and then will its id as finished proposal. So, on next block-build, node will create finishProposalsTx with this proposal.
### AddressStateTx
If BerlinPhase is active, node will reject addressStateTx that tries to edit consortium bit, even if its signed by admin role.

This PR also moves address state definitions from `txs` to their own package, cause its now used from dac pkg as well and this is causing cycle deps.
## How this was tested
Unit-tests, integration test

### CODE REVIEW

File: genesis/camino_genesis.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 	"github.com/ava-labs/avalanchego/vms/avm"
 	"github.com/ava-labs/avalanchego/vms/components/multisig"
 	"github.com/ava-labs/avalanchego/vms/nftfx"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline genesis/camino_genesis.go:23]: small detail but seems not properly formatted. (named import should go at the end of the imports list)

File: vms/platformvm/state/camino_address_state_test.go [hunk 2]
```diff
@@ -11,14 +11,14 @@ import (
 	"github.com/ava-labs/avalanchego/cache"
 	"github.com/ava-labs/avalanchego/database"
 	"github.com/ava-labs/avalanchego/ids"
-	"github.com/ava-labs/avalanchego/vms/platformvm/txs"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/state/camino_address_state_test.go:14]: same here import formatting

File: vms/platformvm/camino_helpers_test.go [hunk 3]
```diff
@@ -27,12 +27,12 @@ import (
 	"github.com/ava-labs/avalanchego/utils/timer/mockable"
 	"github.com/ava-labs/avalanchego/utils/units"
 	"github.com/ava-labs/avalanchego/version"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/camino_helpers_test.go:30]: reorder imports

File: vms/platformvm/camino_service.go [hunk 4]
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/ava-labs/avalanchego/utils/logging"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
 	"github.com/ava-labs/avalanchego/vms/components/keystore"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/camino_service.go:22]: reorder imports

File: vms/platformvm/txs/camino_address_state_tx_test.go [hunk 5]
```diff
@@ -10,6 +10,7 @@ import (
 	"github.com/ava-labs/avalanchego/ids"
 	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/camino_address_state_tx_test.go:13]: reorder imports

File: vms/platformvm/txs/camino_address_state_tx.go [hunk 6]
```diff
@@ -10,66 +10,10 @@ import (
 	"github.com/ava-labs/avalanchego/ids"
 	"github.com/ava-labs/avalanchego/snow"
 	"github.com/ava-labs/avalanchego/vms/components/verify"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/camino_address_state_tx.go:13]: reorder imports

File: vms/platformvm/txs/executor/camino_advance_time_test.go [hunk 7]
```diff
@@ -14,6 +14,7 @@ import (
 	"github.com/ava-labs/avalanchego/utils/constants"
 	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
 	"github.com/ava-labs/avalanchego/utils/nodeid"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/executor/camino_advance_time_test.go:17]: reorder imports

File: vms/platformvm/txs/executor/camino_dac.go [hunk 8]
```diff
@@ -6,6 +6,7 @@ package executor
 import (
 	"errors"
 
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/executor/camino_dac.go:9]: reorder imports

File: vms/platformvm/txs/executor/camino_dac_test.go [hunk 9]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/ava-labs/avalanchego/ids"
 	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/executor/camino_dac_test.go:12]: reorder imports

File: vms/platformvm/txs/executor/camino_tx_executor.go [hunk 10]
```diff
@@ -18,6 +18,8 @@ import (
 	"github.com/ava-labs/avalanchego/vms/components/avax"
 	"github.com/ava-labs/avalanchego/vms/components/multisig"
 	"github.com/ava-labs/avalanchego/vms/components/verify"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/executor/camino_tx_executor.go:21]: reorder imports

File: vms/platformvm/camino_vm_test.go [hunk 11]
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/ava-labs/avalanchego/utils/nodeid"
 	"github.com/ava-labs/avalanchego/utils/set"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/camino_vm_test.go:22]: reorder imports

File: vms/platformvm/dac/camino_add_member_proposal.go [hunk 12]
```diff
@@ -9,13 +9,11 @@ import (
 	"time"
 
 	"github.com/ava-labs/avalanchego/ids"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/dac/camino_add_member_proposal.go:12]: reorder imports

File: vms/platformvm/dac/camino_proposal.go [hunk 13]
```diff
@@ -33,9 +37,19 @@ type Proposal interface {
 
 	StartTime() time.Time
 	EndTime() time.Time
-	GetOptions() any
+	// AddressStateEmpty means that this proposal can't be used as admin proposal
+	AdminProposerAddressState() as.AddressState
```
Comment (knikos) [inline vms/platformvm/dac/camino_proposal.go]: i  would suggest a renaming to sthg like Admin() or AdminRole() or AdminAddressState()
but that's just a suggestion

File: vms/platformvm/dac/camino_base_fee_proposal.go [hunk 14]
```diff
@@ -72,6 +77,23 @@ func (p *BaseFeeProposal) CreateProposalState(allowedVoters []ids.ShortID) Propo
 	return stateProposal
 }
 
+func (p *BaseFeeProposal) CreateFinishedProposalState(optionIndex uint32) (ProposalState, error) {
```
Comment (knikos) [inline vms/platformvm/dac/camino_base_fee_proposal.go:80]: why not call CreateProposalState from within CreateFinsihed.. like this

```go 
func (p *AddMemberProposal) CreateFinishedProposalState(optionIndex uint32) (ProposalState, error) {
	if optionIndex >= 2 {
		return nil, fmt.Errorf("%w (expected: less than 2, actual: %d)", errWrongOptionIndex, optionIndex)
	}
	stateProposal := p.CreateProposalState([]ids.ShortID{}).(*AddMemberProposalState)
	stateProposal.Options[optionIndex].Weight++
	return stateProposal, nil
}
```

File: vms/platformvm/dac/camino_base_fee_proposal.go [hunk 15]
```diff
@@ -10,6 +10,7 @@ import (
 	"time"
 
 	"github.com/ava-labs/avalanchego/ids"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/dac/camino_base_fee_proposal.go:13]: reorder imports

File: vms/platformvm/dac/camino_add_member_proposal.go [hunk 16]
```diff
@@ -67,6 +69,27 @@ func (p *AddMemberProposal) CreateProposalState(allowedVoters []ids.ShortID) Pro
 	return stateProposal
 }
 
+func (p *AddMemberProposal) CreateFinishedProposalState(optionIndex uint32) (ProposalState, error) {
```
Comment (knikos) [inline vms/platformvm/dac/camino_add_member_proposal.go:72]: see BaseFeeProposal suggested change

File: vms/platformvm/state/camino_diff.go [hunk 17]
```diff
@@ -14,11 +14,11 @@ import (
 	"github.com/ava-labs/avalanchego/utils/timer/mockable"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
 	"github.com/ava-labs/avalanchego/vms/components/multisig"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/state/camino_diff.go:17]: reorder imports

File: vms/platformvm/state/camino_diff_test.go [hunk 18]
```diff
@@ -14,11 +14,11 @@ import (
 	"github.com/ava-labs/avalanchego/utils/timer/mockable"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
 	"github.com/ava-labs/avalanchego/vms/components/multisig"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/state/camino_diff_test.go:17]: reorder imports

File: vms/platformvm/state/camino_state.go [hunk 19]
```diff
@@ -11,11 +11,11 @@ import (
 	"github.com/ava-labs/avalanchego/utils/set"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
 	"github.com/ava-labs/avalanchego/vms/components/multisig"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/state/camino_state.go:14]: reorder imports

File: vms/platformvm/state/camino_test.go [hunk 20]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/ava-labs/avalanchego/utils/wrappers"
 	"github.com/ava-labs/avalanchego/version"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/state/camino_test.go:19]: reorder imports

File: vms/platformvm/txs/builder/camino_builder.go [hunk 21]
```diff
@@ -14,6 +14,7 @@ import (
 	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
 	"github.com/ava-labs/avalanchego/utils/timer/mockable"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/builder/camino_builder.go:17]: reorder imports

File: vms/platformvm/txs/builder/camino_builder_test.go [hunk 22]
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/ava-labs/avalanchego/utils/crypto/secp256k1"
 	"github.com/ava-labs/avalanchego/utils/nodeid"
 	"github.com/ava-labs/avalanchego/vms/components/avax"
+	as "github.com/ava-labs/avalanchego/vms/platformvm/addrstate"
```
Comment (knikos) [inline vms/platformvm/txs/builder/camino_builder_test.go:19]: reorder imports

**Review by knikos — COMMENTED** [review]
great job and very goot test coverage!!!  just some comments related mostly to formatting issues.

only one thing to note:  I would add a test that checks whether an Admin Proposal CanBeFinished() immediately after being created just to have that extra unit test.

File: vms/platformvm/dac/camino_base_fee_proposal.go (hunk 14 above)
Comment (evlekht) [reply vms/platformvm/dac/camino_base_fee_proposal.go:80]: cause it needs different TotalAllowedVoters value

File: genesis/camino_genesis.go (hunk 1 above)
Comment (evlekht) [reply genesis/camino_genesis.go:23]: there are already plenty of places where we are importing named pkgs like this
some of them by @peak3d , some by me
our current formatters don't enforce any rule about named imports order
I'd suggest to leave named imports at line where they initially should be, imo
but we have both cases already, so I'd suggest to ignore it until we'll have some formatter that cares about that

Comment (knikos) [reply genesis/camino_genesis.go:23]: yeah actually you are right. they should be enforced by out linter configuration. i will resolve the comments

**Review by knikos — APPROVED** [review]

**Pull request merged**